### PR TITLE
Feature/52 hydromt setup 2d grid

### PR DESCRIPTION
**Pull request by hboisgon**
Add 2D mesh and linked data.
Related issues:

- #52 
- #73 
- #74 
- #75 
- #76

### CODE REVIEW

**Comment by hboisgon** [comment]
Requires branch mesh_methods of hydromt: https://github.com/Deltares/hydromt/pull/186

File: hydrolib/hydromt_delft3dfm/dflowfm.py [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
 import geopandas as gpd
 import hydromt
 import numpy as np
 import pandas as pd
 import pyproj
 import xarray as xr
-from hydromt import gis_utils, io, raster
-from hydromt.models.model_api import Model
+import xugrid as xu
+from hydromt.raster import GEO_MAP_COORD
```
Comment (priscavdsluis) [inline hydrolib/hydromt_delft3dfm/dflowfm.py:17]:
```suggestion
```

File: hydrolib/hydromt_delft3dfm/dflowfm.py [hunk 2]
```diff
@@ -704,12 +694,81 @@ def setup_1dboundary(
...
+        # Function moved to MeshModel in hydromt core
+        # Recreate region dict for core function
+        if mesh2d_fn is not None:
+            region = {"mesh": mesh2d_fn}
+        elif geom_fn is not None:
+            region = {"geom": str(geom_fn)}
+        elif bbox is not None:
+            region = {"bbox": bbox}
+        else:
+            raise ValueError(
```
Comment (priscavdsluis) [inline hydrolib/hydromt_delft3dfm/dflowfm.py:749]: Do we also need to check the validity of the arguments? Such as that bbox should contain 4 values?

File: hydrolib/hydromt_delft3dfm/dflowfm.py [hunk 3]
```diff
@@ -704,12 +694,81 @@ def setup_1dboundary(
...
+        mesh2D_fn : str Path, optional
+            Name of data source for an existing unstructured 2D mesh
+        geom_fn : str Path, optional
+            Path to a polygon used to generate unstructured 2D mesh
+        bbox: list, optional
+            Describing the mesh extent of interest [xmin, ymin, xmax, ymax].
+        res: float, optional
+            Resolution used to generate 2D mesh. By default a value of 100 m is applied.
+
+        See Also
```
Comment (priscavdsluis) [inline hydrolib/hydromt_delft3dfm/dflowfm.py:736]:
```suggestion
            Resolution used to generate 2D mesh. By default a value of 100 m is applied.

    	Raises
        ------
        ValueError
            If `mesh2d_fn`, `geom_fn` and `bbox` are all None.
        IndexError
            If the grid of the spatial domain contains 0 x-coordinates or 0 y-coordinates. 
            
        See Also
```

File: hydrolib/hydromt_delft3dfm/dflowfm.py (hunk 2 above)
Comment (hboisgon) [reply hydrolib/hydromt_delft3dfm/dflowfm.py:748]: Checked in the core super().setup_mesh

File: hydrolib/hydromt_delft3dfm/dflowfm.py [hunk 4]
```diff
@@ -1382,58 +1660,126 @@ def write(self):  # complete model
...
+            elif isinstance(ds, xr.Dataset):
+                for v in ds.data_vars:
+                    if v in self._AUXMAPS:
+                        _prepare_inifields(self._AUXMAPS[v], ds[v])
+                        # update config if frcition
+                        if "frictype" in self._AUXMAPS[name]:
+                            self.set_config(
+                                "physics.UniFrictType", self._AUXMAPS[name]["frictype"]
+                            )
+        # rewrite config
```
Comment (rhutten) [inline hydrolib/hydromt_delft3dfm/dflowfm.py:1746]: Raise value error if both waterlevel and waterdepth are specified self._auxmaps.items().name. Only one can be used.

**Review by rhutten — APPROVED** [review]
Goodwork Hélène

**Pull request merged**